### ISSUE TITLE
Allow adding of custom keys

### ISSUE THREAD

**Issue by snydergd** [dessalines/thumb-key#586]
- [ y] Did you check to see if this issue already exists?
- [y ] Is this only a single feature request? Do not put multiple feature requests in one issue.
- [ n] Is this a question or discussion? Don't use this, use https://lemmy.ml/c/thumbkey

**Describe the feature request below**

I've been really excited about the keyboard as s way to be able to potentially use my phone nearly as efficiently as a regular keyboard and mouse, and have been practicing all weekend on it.

However, one thing that sits in the way of my dream - particularly on the command line, in vim, etc. - is the lack of an escape key and modifiers.

It seems silly to let these couple of keys get in the way, though, and I'm sure that other people have different tweaks they want than me.

Here's what I'm thinking:  Some sort of drawer/screen the user can activate from the main keyboard that could store the extra keys they want.

I would even put some time to a pull request it if the idea sounds palatable.

**Comment by dessalines** [comment]
Rather than doing the complication of adding custom keys, creating space for them, trying to check them in runtime, etc, it'd be much easier to just submit a custom keyboard definition. There are already a few programmer ones that have escape keys and modifiers iirc, but feel free to submit another and I'll be glad to merge it.

**Comment by snydergd** [comment]
I understand that that is easier to implement, at least on a person-by-person basis, but it isn't nearly as good for the user, I don't think, because why would I go through all the work of making a layout that is specific to me (maybe can't even, if not a programmer) and not necessarily creating something someone else will use?  What if I create one with escape, but miss the printscreen key that someone else wanted?  Now they have to create a new layout, too?  Or maybe they just decide this keyboard is not for them and move on.  With a custom key option a variety of people with unique needs can be satisfied even without the perfect layout for them.

**Comment by asdkant** [comment]
I have another use case for custom layouts: I'm a spanish speaker, I used the spanish layout in MessagEase, *with a very small customization*.

Since I moved to thumb-key I've had a hard time with the "messagease spanish" layout, and I'd like to customize it for myself. I'm sure a lot of the user population of this type of keyboard is the type of users who take the time to dig into the config and set things up how they like it, so being able to tune the layout is valuable to the users.

**Comment by dessalines** [comment]
Feel free to create another keyboard for your needs, and I'll gladly add it. But I don't have time to maintain an in-app key customizer.

**Comment by kke** [comment]
> I don't have time to maintain an in-app key customizer.

But you have the time to go through pull requests that add a weird layouts like @asdkant 's "special spanish" with @snyderg 's printscreen mod? And when the list of layouts is 500 items long, some of the them start generating new issues and PRs because someone didn't like them as they were. The people who make the layout PRs also need to learn to compile and run Kotlin. 

Someone else could write the serialization and some simple paste-box (like messagease's hex based keyboard exchange) and maybe even the visual editor. It doesn't have to be you, but you seem to be communicating that you are blocking the addition of this feature and nobody else should contribute on it either. Having so many duplicates of this issue signals that this is something quite many users are wishing for.

I don't know Kotlin, but to me it looks like a basic data structure that should be easily serializable as JSON. If the user provided JSON is silly, then they get a silly keyboard or an error when trying to load it.

**Comment by dessalines** [comment]
It takes only a few minutes to copy-paste someone's file, and have my android compiler tell me that the file is correct. 

> Someone else could write the serialization and some simple paste-box (like messagease's hex based keyboard exchange) and maybe even the visual editor. It doesn't have to be you, but you seem to be communicating that you are blocking the addition of this feature and nobody else should contribute on it either

Having a 2nd format specifically for "non-compile-time-checked" keyboards isn't something I have the time to maintain and fix bugs for. I realize people would like that, as well as an in-app keyboard creator / customizer, but they aren't the ones maintaining this. So many of these keyboards die because the creators don't restrict the feature-set to something they feel they can maintain comfortably.

If someone wants to build a keyboard that works with unchecked keyboard layout files, they can do so.

**Comment by Toldry** [comment]
I've been using MessageEase for a decade and today I discovered ThumbKey. I was really excited when I saw this open source variant exists, and I was ready to switch. 

Now that I see that layout customization is not possible, then I won't make the switch (yet).

I hope this will be reconsidered.

**Comment by dessalines** [comment]
It will not, unless someone wants to take this on, then I'll assign the ticket to you. 

Follow these instructions to add your own layout, like everyone else has done.

**Comment by Toldry** [comment]
@dessalines I am working on implementing this feature.

I already have a working (but very ugly) first draft of the code [here](https://github.com/Toldry/thumb-key/tree/feature/key_modifications).

It works by using a JSON object to represent key modifications.

For example, this JSON applies the following changes to the `Hebrew Message Ease` keyboard:
- KeyboardC[1][1] (center button): 
  - change center click to 𒍅
  - change top right drag to  𒁂
- KeyboardC[2][2] (bottom right button):
  - change bottom right drag to  𒇫
```{
  "HEMessagEase": {
    "main": {
      "arr": [
        [],
        [
          {},
          {
            "center": {
              "text": "𒍅",
              "size": "SMALL"
            },
            "topRight": {
              "text": "𒁂"
            }
          }
        ],
        [
          {},
          {},
          {
            "bottomRight": {
              "text": "𒇫"
            }
          }
        ]
      ]
    }
  }
}
```

Here's what they Hebrew keyboard looks like after the modifications:

![Image](https://github.com/user-attachments/assets/2a037d36-c349-4286-85e2-f4752bf91464)

Once the code is ready I'll send a PR. I hope to finish the rest of the work in 1~2 weeks.

I'll be happy for any feedback if you have any already. Otherwise, wait for the PR and in the meantime please assign this issue to me.

**Comment by dessalines** [comment]
Really depends on how error-checked the code is, and how willing you are to support every feature we have, and every bug that arises. If it doesn't work out, you can always maintain a fork.

**Comment by Toldry** [comment]
I'll do my best to make it error-free and compatible with existing settings (e.g. "hide symbols" and such).

However, I do not intend to implement the entire rich set of the features that can be programmatically configured in the `KeyItemC` class (e.g. setting a custom `KeyAction.ReplaceTrailingWhitespace`). 

Those features go beyond the scope of this issue, and are also beyond the scope of what is currently configurable in the legacy MessagEase. For the purposes of this issue, I expect that a limited subset of the features of `KeyItemC` will suffice for the needs of the vast majority of users.
